### ISSUE TITLE
Support for arbitrary partitioning

### ISSUE THREAD

**Issue by leothomas** [stac-utils/pgstac#83]
The data upload functionality assumes the data is temporally dense and [creates datetime partitions that are 1 week long](https://github.com/stac-utils/pgstac/blob/4b582773f764f317788a8879f5cac3ee4f4f7951/sql/003_items.sql#L133). 

Some datasets (eg: CMIP6) are sparsely populated over a very large period of time (eg: 20 files per month spanning 150 years) - this leads to about 1800 partitions with about 20 files each (the monthly data is timestamped to the 1st of the month, so only the partition for the 1st week of the month is created). 

I haven't benchmarked the search functionality with this many partitions, but after a discussion today with @bitner, I wanted to log an issue for implementing arbitrary partitions, since it seems like a valuable feature anyways.

The STAC API will contain both the CMIP6 dataset and other datasets that are much more temporally dense and with shorter temporal extent. The preferred solution in this case would be to have the ability to make large partitions for the sparsely populated time-ranges and smaller partitions for the densely populated time-ranges (eg: 1 year or 5 year partitions from 1950-1999, week or month partitions from 2000-2030 and then 1 or 5 year partitions again from 2031-2100). 

This can be a function that needs to be executed manually before uploading the data (as opposed to dynamically calculating the partition size during upload). 

A further improvement could be a function to further split up partitions (or even recombine them) in the case of data being added or removed from the STAC database

**Comment by robintw** [comment]
I agree this would be very useful. I just tried to insert some data with some dates that were a long time ago (in the 1800s) and wondered why the insertion seemed to hang for so long. It was actually creating all the partitions between now and the 1800s!

(It was even worse when I accidentally added some data where the date had a year of 9999!)

Some sort of way to configure how the partitions are split, and/or alter them once they've been created would be great.
